### ISSUE TITLE
[It broke]: Webtoons download

### ISSUE THREAD

**Issue by Charles-Em** [C9Glax/tranga#111]
### What is broken?

Hello,

It seems that one of the updates broke the downlaod of webtoons (at least on maga4life)
Problem observed on chapters published on 20/01/2024 and after
Was ok on 13/01/2024
It has been grabbing wrong seasons for new chapters (the first one with keyword index)

For exemple, the gamer chapters 20-29 for season 7 were published, but the chapters I grabbed are the 20-29 of season 2 named as volume 7 with the timestamp from the download of volume 2 (as if it copied them and renamed them)

Other exemple done as a test.
I cleared Tower of God to see what it would do from scratch:
After finishign season 1 of 78 chapters, it is beginning season 2 on its chapter 79 instead of chapter 1 (like before you implemented the webtoon format)

**Comment by C9Glax** [comment]
Hey

So to begin I can't reproduce. I get a 596 chapters of Tower Of God and correct Season/Volume and chapter numbers.

What strikes me as odd:

> (as if it copied them and renamed them)

Tranga doesn't touch downloaded chapter files other than checking if they have been downloaded. There is no renaming or messing with Timestamps either. So I don't know how that would happen.

**Comment by Charles-Em** [comment]
Regarding the renaming, this is not something new, I have noticed when tranga checks for chapter existence and find a match, it can rename the matched file to the corresponding rule. (I noticed that long ago when swapping connectors, if one connector had names for some chapters)
In my case, what happened there is that it renamed volume 1 chapters to volume 7 (didn't notice at first, but the volume 1 chapters 20-29 were missing)
What it means is that it considered the chapters existed under a wrong name.


Now about the current issue, I have removed the docker images and containers and recreated them.
I'm on both versions published on 31/01/2024 (web and api)
It is still not working.
I removed Volume 7 chapters 20-29 to do a test (they were incorrect anyway), and the scan took the ones from season 3 as a match and renamed them to Volume 7 (I already "lost" season 1 and 2 before, meaning it was the first llabeleld 20-29 available)

I added again tower of god from scratch and looked closer at it : 
- manga4life : 595 chapters found
I'll launch it and see what the end result is later.

My supposition is that it correctly checks the number of existing chapters on the website.
But there is a problem when scanning what you currently have.
If my assumption is correct, this is what will happen for tower of god : 
- it will download all 78 chapters of season 1 correctly
- it will then begin season 2, will find a match for all first 78 chapters, renamed them as volume 2 and then begin to download from chapter 79
Because that's what's happening for the other webtoon

**Comment by Charles-Em** [comment]
Ok so it finally reached season 2.

The result is quite funny.

If you just let it go from scratch, it will work => after finishing season 1, it will start season 2 properly and grab the correct chapters.

However, if you cancel and restart, it will start to do some funky shit and rebrand every chapter to the latest season where the same chapter number exists.

For exemple :
I had these : 
Tower of God - Vol.1 Ch.1 .cbz
Tower of God - Vol.2 Ch.1 .cbz

After cancel and restart, I have those : 
Tower of God - Vol.2 Ch.1 .cbz  => it is the correct one
Tower of God - Vol.3 Ch.1 .cbz => it is the Tower of God - Vol.1 Ch.1 .cbz renamed as Vol.3

So all Vol.1 chapters have been renamed as Vol.3 after the restart
And Season 2 has been considered done also up to chapter 78 and is beginning from chatper 79 (even if there is not all named chapters until 78, it did consider them as done)

I'm doing an other test I have in mind and will post the results

**Comment by Charles-Em** [comment]
Of so this is the test I did this time :

I added the serie from scratch : 594 remaining chapters found and prepared as jobs (I don't know why it's not 595 like before, but that's irrelevant to the test, I have noticed that sometimes some chapters are unavailable on the manga websites and they need to fix a link)
I cancelled the job after chapter Tower of God - Vol.1 Ch.1 .cbz was grabbed
I restarted the job, it did the renaming described above (Tower of God - Vol.3 Ch.1 .cbz => it is the Tower of God - Vol.1 Ch.1 .cbz renamed as Vol.3)
And what I expected : 591 remaining chapters found => It considered all chapters 1 as done because of the one from season 1

My guess is that it was like that from the implementation, but I didn't see it because I was doing full jobs from scratch and in this case it works. (If you don't cancel it, it will prepare all chapter jobs and execute them)
I only noticed it because of new chapters for the gamer.


TLDR : it works if you downlaod from scratch
But if you cancel and relaunch, it will flag as done all chapters for all seasons if the chapter exists in any season.

**Comment by C9Glax** [comment]
Ugh I found the issue, when checking if chapters exist I only checked for the chapterNumber... Relic of one of the first versions..

**Comment by Charles-Em** [comment]
I swapped on cuttingedge to test, but an other change (the trailing spaces) made my library go crazy on some mangas so I cancelled the check and went back to latest currently

**Comment by C9Glax** [comment]
Oh no I messed up the regex...

**Comment by C9Glax** [comment]
please dont try again for now still working on it

**Comment by C9Glax** [comment]
Okay fixed it. You may try again :)

**Comment by Charles-Em** [comment]
Ok first feedback : 
- regarding the regex issue, out of 41 mangas, 39 are ok
- for the 2 others, he is trying to redownload them fully, currently put them "on-hold" and will ook into it later

Then regarding ths current topic : 
- reactivating tower of god : looks ok regarding scanning the current status (correctly mapping the missing chapters)
- I downloaded the one chapter which was missing from season 2 (dunno why), then cancelled, then relaunched. And the program still thinks its missing. But since it was downloaded, he is stuck not doing anything 
[02/07/2024 17:40:57.388]     MangaLife | Downloading Images for /Manga/Tower of God/Tower of God - Vol.2 Ch.55.cbz

Reactivating the gamer which was full cause I managed to do a full redownload in one pass
Taking into account the following info : 
- The first season is 86 chapters long
- The only other season longer is season 4 with 154 chapters
=> The program is saying chapters 87 to 154 are missing from first season ans is currently downloading season 4 chapters 87-154 as season 1

For later tests I will host an oterh container for cutting edge API to not mess up with my live settings

**Comment by Charles-Em** [comment]
Ok looked into the 2 mangas he wants to full redownload : 

Current chapters are like this "Chainsaw Man - Vol.1 Ch.1 .cbz"
And he wants to redo them like this "Chainsaw Man - Vol.1 Ch.1.cbz"

Removing a space.

But I have a connected kavita library, refresh my memory, what information is the program getting from kavita? (although I dont have any GET librabry in the logs anyway after restart)

**Comment by C9Glax** [comment]
> Reactivating the gamer which was full cause I managed to do a full redownload in one pass

Yes as said before the whole logic of checking which chapters already exist was flawed, so on the first pass (when there were obviously no existing chapters) it would work correctly. But after that it would only compare chapter numbers (so vol1. ch1 and vol2. ch1. would return both as existing if any chapter 1 was existing). Fixed that.


> Ok looked into the 2 mangas he wants to full redownload 
> 
> Current chapters are like this "Chainsaw Man - Vol.1 Ch.1 .cbz"
> And he wants to redo them like this "Chainsaw Man - Vol.1 Ch.1.cbz"
> 
> Removing a space.

It's doing this because it didn'T think the chapters where downloaded... Regex matching is weird, but I made it work for those too now...

> what information is the program getting from kavita?

It's only retrieving library-IDs and then calling a fresh on those.

**Comment by Charles-Em** [comment]
> > Reactivating the gamer which was full cause I managed to do a full redownload in one pass
> 
> Yes as said before the whole logic of checking which chapters already exist was flawed, so on the first pass (when there were obviously no existing chapters) it would work correctly. But after that it would only compare chapter numbers (so vol1. ch1 and vol2. ch1. would return both as existing if any chapter 1 was existing). Fixed that.

Yeah, but I figured that I should do a test on a full webcomic to validate it.
And as described there is an issue where he wants to "fake complete" season 1 with season 4 chapters (87 to 154)
This issue is only specific to season 1 and is not happening on tower of god (I have no hypothesis on the underlying issue as a result)
To be noted that he isn't removing anything doing that.

> > Ok looked into the 2 mangas he wants to full redownload
> > Current chapters are like this "Chainsaw Man - Vol.1 Ch.1 .cbz"
> > And he wants to redo them like this "Chainsaw Man - Vol.1 Ch.1.cbz"
> > Removing a space.
> 
> It's doing this because it didn'T think the chapters where downloaded... Regex matching is weird, but I made it work for those too now...

I mentionned an other new issue, I will describe it properly because I kind of had a hard time doing it.
Let's say I add a new manga which has 3 chapters.
Step 1 : the 3 chapters get downloaded properly (no issue)
Step 2 : I cancel/relaunch the manga or restart the container
Step 3 : the program will retry to download the 3 chapters (first issue is thinking they are missing)
Step 4 : it times out at 5 minutes on each chapter instead of redownloading it (probably because they already exist in this case)

Almost posted there, but I found something interesting.

Your latest fix corrected the check on chapter existence when there is a blank space before the ".cbz" but not if there is none
If my chapter is named "Chainsaw Man - Vol.1 Ch.1.cbz" it get stuck 5 minutes
If my chapter is named "Chainsaw Man - Vol.1 Ch.1 .cbz", it is detected as existing and pass it

**Comment by Charles-Em** [comment]
Ok this part is fixed :
> If my chapter is named "Chainsaw Man - Vol.1 Ch.1.cbz" it get stuck 5 minutes
> If my chapter is named "Chainsaw Man - Vol.1 Ch.1 .cbz", it is detected as existing and pass it
I activated cuttingedge on my live manga list, and there was no manga that had any issue trying to redownload everything



Then about the "it is fixed" on the gamer, I retract it half (lol)

There is no problem if I get it from Manga4life : 496 chapters retrieved which is correct and 0 new publications found (I have them all)
There is a problem if I get it from Mangasee : 496 chapters retrieved which is correct, but then matching 68 new chapters where it is trying to label season 4 87-154 into season 1 (season 1 ends at 86)

[02/07/2024 18:46:15.581]      Mangasee | Got 496 chapters. Publication The Gamer dGhlZ2FtZXIyMDEz
[02/07/2024 18:46:15.586]      Mangasee | Checking for duplicates Publication The Gamer dGhlZ2FtZXIyMDEz
[02/07/2024 18:46:16.833]      Mangasee | 68 new chapters. Publication The Gamer dGhlZ2FtZXIyMDEz

[02/07/2024 18:48:21.369]     MangaLife | Got 496 chapters. Publication The Gamer dGhlZ2FtZXIyMDEz
[02/07/2024 18:48:21.376]     MangaLife | Checking for duplicates Publication The Gamer dGhlZ2FtZXIyMDEz
[02/07/2024 18:48:22.776]     MangaLife | 0 new chapters. Publication The Gamer dGhlZ2FtZXIyMDEz

Connector issue ? I will get them from mangalife at the moment.

PS : no problem, I wish I could code to help there lol. I have only been manipulating data for the last 10 years, So apart from SQL and algorithms, I can't do much
Have to stop there this evening though.

**Comment by Charles-Em** [comment]
Hmm the issue of trying to redownload and timing out on 5 minutes because of the last space is still doing it on ToG
"Tower of God - Vol.3 Ch.51.cbz" stuck
"Tower of God - Vol.3 Ch.51 .cbz" pass

**Comment by C9Glax** [comment]
If it's okay please close again 👍🏼

**Comment by Charles-Em** [comment]
There is still this part remaining :


There is a problem if I get it from Mangasee : 496 chapters retrieved which is correct, but then matching 68 new chapters where it is trying to label season 4 87-154 into season 1 (season 1 ends at 86)

[02/07/2024 18:46:15.581] Mangasee | Got 496 chapters. Publication The Gamer dGhlZ2FtZXIyMDEz
[02/07/2024 18:46:15.586] Mangasee | Checking for duplicates Publication The Gamer dGhlZ2FtZXIyMDEz
[02/07/2024 18:46:16.833] Mangasee | 68 new chapters. Publication The Gamer dGhlZ2FtZXIyMDEz

[02/07/2024 18:48:21.369] MangaLife | Got 496 chapters. Publication The Gamer dGhlZ2FtZXIyMDEz
[02/07/2024 18:48:21.376] MangaLife | Checking for duplicates Publication The Gamer dGhlZ2FtZXIyMDEz
[02/07/2024 18:48:22.776] MangaLife | 0 new chapters. Publication The Gamer dGhlZ2FtZXIyMDEz

This is more about the connector imo.

**Comment by C9Glax** [comment]
Yea the connector for whatever reason always set volumeNumber to 1.
don't ask me why i did what i did.

**Comment by Charles-Em** [comment]
I haven't updated since then seeing all those new issues popping up 🗡️ 
Is it safe to update? Which version? Then I'll test

**Comment by C9Glax** [comment]
If you want to be on the safe side right now dont update yet.
Everything is still running, but behaviour rn is kinda weird.

**Comment by C9Glax** [comment]
`cuttingedge` should be good to run for now if you were still waiting...
Gonna close this as I don't see anything still being bad
